lib/NixFile make absolute by construction

Profpatsch opened the pull request:
This expands on https://github.com/target/lorri/pull/194 a bit by making `NixFile` always be an absolute path, correct by construction.

It then factors out the absolute path part to also add it to `cas.rs` and `constants/Paths`.
We should propagate further into `project.rs` and `roots.rs`, but I’ve left that as a TODO for now.

Most of these changes are rather mechanical (and mostly trivial).

cc @curiousleo

curiousleo commented on src/constants.rs:
```diff
@@ -18,33 +18,49 @@ impl Paths {
+        // TODO: return as good error value
+        assert!(
+            pd.cache_dir().is_absolute(),
+            "Your cache directory is not an absolute path! It is: {}",
```
Is there anything the user could do about this error message when they see it? `pd` comes from `ProjectDirs::from`, so it seems to me like that method should ensure this is a "good" path. Perhaps `ProjectDirs::cache_dir` could return an `AbsPathBuf`?

Profpatsch replied:
`ProjectDirs::cache_dir` comes from a different library. I opted for a `panic!()` for now, because if the paths return by `ProjectDirs` are not absolute, we have a problem (because we want to ensure that nix files are absolute).

curiousleo commented on src/constants.rs:
```diff
@@ -18,33 +18,49 @@ impl Paths {
+        // TODO: return as good error value
+        assert!(
+            runtime_dir.is_absolute(),
+            "Your runtime directory is not an absolute path! It is: {}",
```
Same as above

curiousleo commented on src/main.rs:
```diff
@@ -44,48 +44,56 @@ fn main() {
-    Ok(NixFile::from(locate_file::in_cwd(&shellfile).map_err(
-        |_| {
+    Ok(NixFile::from(
+        // `::in_cwd` is guaranteed to return an absolute path
```
Shouldn't `in_cwd` just return an `AbsPathBuf` in that case?

Profpatsch replied:
Good idea, changed in last commit.

Profpatsch commented:
> I don't understand this PR

We must ensure that `.nix` files are always absolute. Otherwise the daemon cannot function correctly. While adding that guarantee I noticed that we should guarantee it for the `cas` module as well, because we generate nix paths from that, which are read by the daemon. So I factored the shared preconditions into `AbsPathBuf`.
This follows the approach of [parse, don’t validate](https://lexi-lambda.github.io/blog/2019/11/05/parse-don-t-validate/) (actually this commit is from before that blogpost, but that explains the ideas quite well. An older and more concise description is in [Matt Parson’s blog](https://www.parsonsmatt.org/2017/10/11/type_safety_back_and_forth.html).

> the only way to construct one is via `new_unchecked`. This means there is no actual verification "by construction" that the contained PathBuf is actually absolute.

I added a `new` function which does a check, and added it at the borders where we cannot be entirely sure that a path is absolute.

Profpatsch commented:
rebased on master.

gilligan commented:
@grahamc you had blocked this a good while ago. Does your opinion still hold? Should this be abandoned or is there interest on either side to pick this up again?

Profpatsch commented:
I’m sure it’s going to prevent bugs in the future, but my mental health does not depend on it being merged :)

Profpatsch commented:
Merged in https://github.com/nix-community/lorri/pull/14
